### PR TITLE
Fix #17229 - Refactor function that exports bin info add tests and fix segfault ##bin

### LINKED ISSUE #17229: Refactor r_core_bin_export_info_rad to accept a mode
Current code for `r_core_bin_export_info_rad`: https://github.com/radareorg/radare2/blob/08e52d834ae0d840b97fde94186c0615bb83d333/libr/core/cbin.c#L3982

This function is used in https://github.com/radareorg/radare2/blob/08e52d834ae0d840b97fde94186c0615bb83d333/libr/core/cbin.c#L176 to get its output and then execute it. However, it would be better to directly have `r_core_bin_export_info_rad` set the right things.

The new signature should be something like `R_API void r_core_bin_export_info_rad(RCore *core, int mode)` and mode can be `R_MODE_PRINT`, `R_MODE_JSON`, `R_MODE_SET`, `R_MODE_RADARE`, etc.

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**

- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

@ret2libc i dont think we have any test for this either, did you verified that before merging your PR?

**Pending stuff**

* [x] Do not use r_core_cmd at all
* [x] Implement proper clean C APIs to achieve those commands without having to type 40 locs
* [x] Support quiet mode (not sure why :?)
* [x] Avoid the use of strings as much as possible

**Test plan**

there must be some tests at least to verify the behaviour of -r and mode.set are the same

**Closing issues**
#17229

### CODE REVIEW

**Comment by trufae** [comment]
The title say "WIP", the description contains a list of items to not be wip, some tests are needed to verify the changes are not breaking anything, you didn't added tests in your PR, so this one aims to fix all the issues after the merge.

Also, some core_cmd's are still called in here, and the way to replace in C requires more than 20 LOC, so imho this is wrong and i may think on proper APIs to expose this functionality in here.

File: libr/core/cmd_print.c [hunk 1]
```diff
@@ -1528,12 +1527,11 @@ static void cmd_print_format(RCore *core, const char *_input, const ut8* block,
...
 			/* store a new format */
 			if (space && (!eq || space < eq)) {
-				// char *fields = NULL;
 				*space++ = 0;
-				// fields = strchr (space, ' ');
-				if (strchr (name, '.')) {// || (fields != NULL && strchr(fields, '.') != NULL)) // if anon struct, then field can have '.'
+				if (strchr (name, '.')) {
 					eprintf ("Struct or fields name can not contain dot symbol (.)\n");
 				} else {
+					// pf.foo=xxx
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd_print.c:1534]: why? it helps to understand which command is needed to run this code. actually the whole function is an antipattern and must be refactored quite heavily

[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd_print.c:1534]: but out of the scope of this pr

File: libr/core/cmd_open.c [hunk 2]
```diff
@@ -1690,22 +1690,24 @@ static int cmd_open(void *data, const char *input) {
...
+				RIODesc *desc = r_io_desc_get (core->io, core->file->fd);
 				perms = (input[3] == '+')? R_PERM_R|R_PERM_W: 0;
-				r_core_file_reopen (core, input + 4, perms, 0);
 				if (desc) {
-					r_core_bin_load_structs (core, desc->name);
+					char *fname = strdup (desc->name);
+					r_core_file_reopen (core, input + 4, perms, 0);
+					r_core_bin_load_structs (core, fname);
+					free (fname);
 				}
```
[the remark this answers is not in the document]
Comment (trufae) [reply libr/core/cmd_open.c:1705]: uhm, first i though nope. but i think it should pushing and waiting for travis

File: libr/core/cmd_open.c [hunk 3]
```diff
@@ -1690,24 +1690,35 @@ static int cmd_open(void *data, const char *input) {
...
 			break;
 		case 'n': // "oon"
-			if ('n' == input[2]) {
-				RIODesc *desc = r_io_desc_get (core->io, core->file->fd);
-				if ('?' == input[3]) {
+			switch (input[2]) {
+			case 0: // "oon"
+				r_core_file_reopen (core, NULL, 0, 0);
+				break;
+			case ' ': //"oon file"
```
Comment (ret2libc) [inline libr/core/cmd_open.c:1697]: add this to the help msg

Comment (trufae) [reply libr/core/cmd_open.c:1697]: Actually this thing is not even working, so its probably better not to document it, and work on fixing and adding tests for this in a separate PR.

File: libr/core/cmd_open.c [hunk 4]
```diff
@@ -1679,24 +1683,36 @@ static int cmd_open(void *data, const char *input) {
...
-				r_core_file_reopen (core, input + 4, perms, 0);
+				RIODesc *desc = r_io_desc_get (core->io, core->file->fd);
 				if (desc) {
-					r_core_bin_load_structs (core, desc->name);
+					perms = core->io->desc->perm;
+					if (input[3] == '+') {
+						perms |= R_PERM_RW;
+					}
+					char *fname = strdup (desc->name);
+					r_core_bin_load_structs (core, fname);
```
Comment (ret2libc) [inline libr/core/cmd_open.c:1705]: as said on Telegram, I'm not sure why we do load_structs and then file_reopen.

Comment (trufae) [reply libr/core/cmd_open.c:1705]: i fixed it.. but github doesnt seems to reflect the last commit in the pr.. only in the branch wtf

**Comment by XVilka** [comment]
```


[XX] db/cmd/cmd_open oonn
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'pf.~?
oonn
pf.~?
o~?
' bins/mach0/mac-ls2
-- stdout
@@ -1,3 +1,3 @@
 0
 58
-1
+2


-- stderr
File bins/mach0/mac-ls2 reopened in read-only mode
```
Also these are on OpenBSD:
```
[XX] db/formats/mdmp : test format definitions
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -nn -Qc pf. bins/mdmp/hello.dmp
-- stdout
@@ -1,21 +0,0 @@
-pf.mdmp_directory [4]E? (mdmp_stream_type)StreamType (mdmp_location_descriptor)Location
-pf.mdmp_exception [4]E[4]Eqqdd[15]q (mdmp_exception_code)ExceptionCode (mdmp_exception_flags)ExceptionFlags ExceptionRecord ExceptionAddress NumberParameters __UnusedAlignment ExceptionInformation
-pf.mdmp_exception_stream dd?? ThreadId __Alignment (mdmp_exception)ExceptionRecord (mdmp_location_descriptor)ThreadContext
-pf.mdmp_header [4]zddddt[8]B Signature Version NumberOfStreams StreamDirectoryRVA CheckSum TimeDateStamp (mdmp_type)Flags
-pf.mdmp_location_descriptor dd DataSize RVA
-pf.mdmp_location_descriptor64 qq DataSize RVA
-pf.mdmp_memory64_list qq[83]? NumberOfMemoryRanges BaseRva (mdmp_memory_descriptor64)MemoryRanges
-pf.mdmp_memory_descriptor q? StartOfMemoryRange (mdmp_location_descriptor)Memory
-pf.mdmp_memory_descriptor64 qq StartOfMemoryRange DataSize
-pf.mdmp_memory_info qq[4]Edq[4]E[4]E[4]Ed BaseAddress AllocationBase (mdmp_page_protect)AllocationProtect __Alignment1 RegionSize (mdmp_mem_state)State (mdmp_page_protect)Protect (mdmp_mem_type)Type __Alignment2
-pf.mdmp_memory_info_list ddq[127]? SizeOfHeader SizeOfEntry NumberOfEntries (mdmp_memory_info)MemoryInfo
-pf.mdmp_misc_info d[4]Bdtttddddd SizeOfInfo (mdmp_misc1_flags)Flags1 ProcessId ProcessCreateTime ProcessUserTime ProcessKernelTime ProcessorMaxMhz ProcessorCurrentMhz ProcessorMhzLimit ProcessorMaxIdleState ProcessorCurrentIdleState
-pf.mdmp_module qddtd???qq BaseOfImage SizeOfImage CheckSum TimeDateStamp ModuleNameRVA (mdmp_vs_fixedfileinfo)VersionInfo (mdmp_location_descriptor)CvRecord (mdmp_location_descriptor)MiscRecord Reserved0 Reserved1
-pf.mdmp_module_list d[10]? NumberOfModule (mdmp_module)Modules
-pf.mdmp_string dZ Length Buffer
-pf.mdmp_system_info [2]EwwbBddd[4]Ed[2]Ew[2]q (mdmp_processor_architecture)ProcessorArchitecture ProcessorLevel ProcessorRevision NumberOfProcessors (mdmp_product_type)ProductType MajorVersion MinorVersion BuildNumber (mdmp_platform_id)PlatformId CsdVersionRva (mdmp_suite_mask)SuiteMask Reserved2 ProcessorFeatures
-pf.mdmp_thread ddddq?? ThreadId SuspendCount PriorityClass Priority Teb (mdmp_memory_descriptor)Stack (mdmp_location_descriptor)ThreadContext
-pf.mdmp_thread_list d[1]? NumberOfThreads (mdmp_thread)Threads
-pf.mdmp_token_info ddq TokenSize TokenId TokenHandle
-pf.mdmp_token_info_list dddd TokenListSize TokenListEntries ListHeaderSize ElementHeaderSize
-pf.mdmp_vs_fixedfileinfo ddddddddddddd dwSignature dwStrucVersion dwFileVersionMs dwFileVersionLs dwProductVersionMs dwProductVersionLs dwFileFlagsMask dwFileFlags dwFileOs dwFileType dwFileSubtype dwFileDateMs dwFileDateLs


-- stderr
[WARN] Invalid or unsupported enumeration encountered 21
[WARN] Invalid or unsupported enumeration encountered 22
Error parsing resource directory
[INFO] Parsing data sections for large dumps can take time, please be patient (but if strings ain't your thing try with -z)!

-- exit status: 
[XX] db/cmd/feat_grep escaped chars
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?e Hello\x40World~Hello\@
?e Hello\x24World~Hello\@
?e Hello\n\x23World~\#
' -
-- stdout
@@ -1,2 +1 @@
 Hello@World
-#World


-- exit status: 
[XX] db/cmd/cmd_open oonn
R2_NOPLUGINS=1 radare2 -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'pf.~?
oonn
pf.~?
o~?
' bins/mach0/mac-ls2
-- stdout
@@ -1,3 +1,3 @@
 0
 58
-1
+2


-- stderr
File bins/mach0/mac-ls2 reopened in read-only mode
```

**Comment by trufae** [comment]
@XVilka are those issues related to this PR?

**Comment by trufae** [comment]
So, as i said swapping the lines caused mayhem. swapped them back. no room to investigate this in this PR imho. We have more tests and a native implementation without running external or internal commands now. so it's a considerable improvement.

**Comment by trufae** [comment]
OpenBSD is still red for a reason. i would prefer to find out why before mergingh this and gettnig master in red

**Comment by trufae** [comment]
probably related to those asans

Screenshot 2020-07-17 at 16 33 47

**Comment by trufae** [comment]
Screenshot 2020-07-17 at 16 37 48

**Comment by trufae** [comment]
Yay!!1

**Pull request merged**